File: mobile/src/features/scan-context/prefill.ts
/**
 * Filling the context form from what the label says — FR-03's "pre-filled from OCR".
 *
 * The form exists because the rules engine needs a product profile, and three of its fields decide
 * which rules run at all (`profile.ts`). That is a good reason for the *data* to exist and a bad
 * reason to make a person type it after every photograph, when the pack in their hand already says
 * most of it. So the backend reads the photograph and proposes values; this module decides what the
 * form does with the proposals.
 *
 * **Suggestions fill blanks. They never overwrite a person.** `applySuggestions` takes the current
 * form values and returns only the fields that were empty. A prefill that lands while someone is
 * halfway through typing a quantity must not move it under their thumb, and a user who corrects a
 * suggested value must not have the correction reverted by a late poll.
 *
 * **Nothing here is accepted silently.** Every field this fills comes back in `filled`, which the
 * screen uses to mark the field as read-from-the-label and to gate submission behind one
 * confirmation. The machine may fill the form; the person still affirms it before a scan exists
 * (CLAUDE.md §3.1). `RULE_RELEVANT` names the fields where that gate is not optional.
 *
 * **The category is matched here, not on the server.** The 26-code vocabulary is the client's
 * (`categories.ts`) and the server has no copy of it — duplicating the list in Python would be two
 * lists to keep in step and a code that could drift out of a rule pack's filter. The server
 * proposes the commodity's common name; `suggestCategory` matches that against the keywords the
 * search box already uses. A name that matches two categories equally well matches neither: an
 * ambiguous guess here is a wrong `category_code` on a BIS applicability lookup.
 */

import type { TranslationKey } from '@/i18n';

import { CATEGORIES, type Category } from './categories';
import type { ContextFormValues } from './profile';

/** One proposed value, as the server sends it. */
export interface Suggestion {
  /** A profile field name: `name`, `net_qty_value`, `net_qty_unit` or `is_imported`. */
  field: string;
  /** Always a string, including for the boolean (`'true'`). */
  value: string;
  /** Inherited from the declaration it was read from. Below 0.75 the machine is unsure. */
  confidence: number;
  /** The declaration it came from — shown so "why does it say this" has an answer. */
  fromFieldCode: string;
  /** What was recognised on the pack, verbatim. */
  sourceText: string;
}

/** Where one prefill has got to. */
export type PrefillStatus = 'reading' | 'ready' | 'failed';

export interface PrefillResult {
  prefillId: string;
  status: PrefillStatus;
  suggestions: Suggestion[];
  /** Words recognised. Zero with no suggestions means the photograph was read and had nothing on it. */
  wordCount: number;
  /** Extraction ran pattern-only, so no product name was proposed. */
  reduced: boolean;
}

/**
 * FR-06's threshold, repeated here rather than imported from the findings feature.
 *
 * The same number for the same reason — below it the machine does not believe its own reading — but
 * a different decision hangs on it: there, whether a verdict may rest on a value; here, only how
 * loudly a filled field says "check this". Coupling the two would make a change to one a change to
 * the other.
 */
export const UNSURE_BELOW = 0.75;

/**
 * Form fields that change which rules run, and therefore cannot be accepted without a person.
 *
 * Surface is the third rule-relevant field and is deliberately absent: the server never proposes
 * it (nothing in a label's words says whether its text is embossed), so it is never in a state that
 * needs confirming.
 */
export const RULE_RELEVANT: readonly (keyof ContextFormValues)[] = [
  'quantityValue',
  'quantityUnit',
  'isImported',
];

/** The form field each suggested profile field lands in. */
const FIELD_FOR: Readonly<Record<string, keyof ContextFormValues>> = {
  name: 'name',
  net_qty_value: 'quantityValue',
  net_qty_unit: 'quantityUnit',
  is_imported: 'isImported',
};

/** Whether a form value is empty enough to be filled without overwriting anyone. */
function isBlank(value: unknown): boolean {
  if (typeof value === 'string') return value.trim().length === 0;
  // `isImported` is a boolean with a default of false. False is "not yet said", which is why the
  // server never proposes a `false` — see the note on `applySuggestions`.
  return value === false || value === undefined || value === null;
}

export interface AppliedPrefill {
  /** Only the fields that were blank. Spread over the form; never a whole form. */
  values: Partial<ContextFormValues>;
  /** What filled each one, keyed by form field — the screen's provenance chips. */
  filled: Partial<Record<keyof ContextFormValues, Suggestion>>;
}

/**
 * Work out which form fields a set of suggestions may fill.
 *
 * Args:
 *   current: the form as it stands. A non-blank field is left alone.
 *   suggestions: what the server read off the label.
 *
 * Returns the values to apply and the provenance for each.
 *
 * **`isImported` is only ever set to true.** The server proposes it in one direction for a reason
 * that matters more than it looks: `is_imported=false` switches the importer and country-of-origin
 * rules *off*, and a pack with no importer line is exactly the pack most likely to be in breach of
 * Rule 6. So an absent declaration never becomes a filled `domestic`, here or there — the form's
 * own default stands, where the user can see it and change it.
 */
export function applySuggestions(
  current: ContextFormValues,
  suggestions: readonly Suggestion[],
  translate: (key: TranslationKey) => string
): AppliedPrefill {
  const values: Partial<ContextFormValues> = {};
  const filled: Partial<Record<keyof ContextFormValues, Suggestion>> = {};

  for (const suggestion of suggestions) {
    const field = FIELD_FOR[suggestion.field];
    if (!field) continue; // a field this build does not know — ignored, never guessed at
    if (!isBlank(current[field])) continue;

    if (field === 'isImported') {
      if (suggestion.value !== 'true') continue;
      values.isImported = true;
    } else {
      const text = suggestion.value.trim();
      if (text.length === 0) continue;
      (values as Record<string, string>)[field] = text;
    }

    filled[field] = suggestion;
  }

  // The category follows from the name rather than from a declaration of its own, so it is derived
  // here — after the loop, from whichever name is now in play — rather than proposed by the server,
  // which has no copy of the vocabulary (see the module docstring).
  const name = values.name ?? current.name;
  if (name && isBlank(current.categoryCode)) {
    const code = suggestCategory(name, translate);
    if (code) {
      values.categoryCode = code;
      filled.categoryCode = {
        field: 'category_code',
        value: code,
        // Inherited from the name it was derived from: a category matched off an uncertain reading
        // is no more certain than the reading.
        confidence: filled.name?.confidence ?? 1,
        fromFieldCode: filled.name?.fromFieldCode ?? 'common_name',
        sourceText: name,
      };
    }
  }

  return { values, filled };
}

/**
 * The category a product name implies, or null when the name does not clearly say.
 *
 * Scored rather than filtered: `searchCategories` answers "what could the user mean" for a search
 * box, where showing six options is helpful. This answers "which one is it" for a field that goes
 * to a BIS applicability lookup, where a wrong code is a wrong answer about whether a licence is
 * needed. So a tie wins nothing.
 *
 * `translate` is passed in rather than imported, like `searchCategories`, so the match works in
 * either locale and the function stays pure.
 */
export function suggestCategory(
  name: string,
  translate: (key: TranslationKey) => string
): string | null {
  const haystack = name.toLowerCase();
  if (haystack.trim().length === 0) return null;

  const scored = CATEGORIES.map((category) => ({
    category,
    score: scoreCategory(category, haystack, translate),
  })).filter((entry) => entry.score > 0);

  if (scored.length === 0) return null;

  const best = Math.max(...scored.map((entry) => entry.score));
  const winners = scored.filter((entry) => entry.score === best);

  // A tie is ambiguity, not a coin toss. "Mustard Oil" scoring equally as oil and as spices means
  // the label did not say, and the user picks from the list they would have used anyway.
  return winners.length === 1 ? winners[0].category.code : null;
}

/**
 * How well one category matches a product name.
 *
 * A label match counts double: "Edible Oil" naming the category outright is stronger evidence than
 * a keyword that happens to appear inside a longer word.
 */
function scoreCategory(
  category: Category,
  haystack: string,
  translate: (key: TranslationKey) => string
): number {
  let score = 0;

  const label = translate(category.labelKey).toLowerCase();
  if (label.length > 0 && haystack.includes(label)) score += 2;

  for (const keyword of category.keywords) {
    if (matchesWord(haystack, keyword)) score += 1;
  }

  return score;
}

/**
 * Whether a keyword appears in the name as a word rather than as a fragment.
 *
 * Fragments produce nonsense: "oil" sits inside "boiled", and matching it would file boiled sweets
 * under edible oils. So the character before the keyword must not be part of a word.
 *
 * **A plural suffix is still the word.** Packs say "Biscuits", "Sweets", "Pulses" — requiring an
 * exact boundary on both sides makes the keyword list miss the form the words are actually printed
 * in, which is the only form that matters here. A trailing `s` or `es` is allowed; anything longer
 * is a different word.
 *
 * Checked by hand rather than with a `RegExp` built from the keyword, which would need escaping for
 * a list that is fixed, small and inspectable.
 */
function matchesWord(haystack: string, keyword: string): boolean {
  let from = 0;
  for (;;) {
    const at = haystack.indexOf(keyword, from);
    if (at < 0) return false;

    const before = at === 0 ? ' ' : haystack[at - 1];
    if (!isWordChar(before) && endsWord(haystack, at + keyword.length)) return true;

    from = at + 1;
  }
}

/** Whether the word ends at `at`, allowing an English plural suffix. */
function endsWord(haystack: string, at: number): boolean {
  const charAt = (index: number) => (index >= haystack.length ? ' ' : haystack[index]);

  if (!isWordChar(charAt(at))) return true;
  if (charAt(at) === 's' && !isWordChar(charAt(at + 1))) return true;
  return charAt(at) === 'e' && charAt(at + 1) === 's' && !isWordChar(charAt(at + 2));
}

function isWordChar(char: string): boolean {
  return /[a-z0-9]/.test(char);
}

/**
 * Which rule-relevant fields a prefill filled and a person has not yet touched.
 *
 * The screen's submit gate reads this: while it is non-empty, the form has machine-supplied values
 * behind a rule that decides whether a rule runs, and the user is asked to confirm them once. It
 * empties when they confirm, and also when they edit the field themselves — editing *is*
 * confirming, and asking twice for the same assurance teaches people to tap past it.
 */
export function unconfirmedRuleFields(
  filled: Partial<Record<keyof ContextFormValues, Suggestion>>,
  confirmed: boolean
): (keyof ContextFormValues)[] {
  if (confirmed) return [];
  return RULE_RELEVANT.filter((field) => filled[field] !== undefined);
}

/** Whether any field was filled at all — what the screen's banner is about. */
export function anyFilled(filled: Partial<Record<keyof ContextFormValues, Suggestion>>): boolean {
  return Object.keys(filled).length > 0;
}

File: backend/app/services/extraction/prefill.py
"""Turning read declarations into a product-context *suggestion* — FR-03's prefill.

FR-03 asks for a product context form and then adds: "Three fields are pre-filled from OCR ...
and confirmed by the user." This module is the "pre-filled from OCR" half. It takes the
declarations extraction already found on a label and proposes form values for them.

**It suggests. It never decides.** Nothing here produces a verdict, and nothing here writes a
profile. The output is a list of proposals that a person accepts, edits or ignores before a scan
is created, which is the only reason using a model on this path is acceptable at all: the profile
that reaches ``evaluate()`` is still the one a human affirmed (CLAUDE.md §3.1).

Three rules keep that honest, and each of them is a safety property rather than a style choice.

**1. A suggestion is only ever made from a declaration that was actually read.** Every proposal
carries the ``field_code`` it came from and the text that was recognised, so "why is the form
saying 500 g" always has an answer that points at the pack. An inference from *absence* is not a
reading, and is never made — see rule 2.

**2. Prefill may turn importer rules ON. It may never turn them off.** ``is_imported`` decides
whether the importer and country-of-origin rules run at all, so a wrong ``false`` does not produce
a wrong FAIL — it produces silence, which is worse, because the pack whose importer line is
missing is exactly the pack that reads as domestic. So an importer declaration on the label
suggests *imported*; the absence of one suggests **nothing**, and the form's own default stands
where the user can see and change it.

**3. Surface is never suggested.** Rule 9 gives embossed, blown, moulded and perforated text a
higher threshold than printed text, and which of those a pack is cannot be read from its words —
it is a property of the physical surface. A guess here moves a threshold column, so there is no
guess. ``NEVER_SUGGESTED`` records this and the test suite asserts it.

**What is deliberately not here:** the category code. The category vocabulary is the client's
(``mobile/src/features/scan-context/categories.ts``) and duplicating those 26 codes in a ``.py``
file would create two lists to keep in step for no gain. This module proposes the *name*; the
client matches it against the vocabulary it already owns.

Pure, like ``rules/evaluate()``: no I/O, no model call, no clock. The model ran earlier, in
``extraction.llm_layer``, and its output arrives here as ordinary ``Extraction`` values — so no
new LLM call site is introduced and §9's table still lists three.
"""

from __future__ import annotations

import re
from collections.abc import Sequence
from dataclasses import dataclass

from app.services.extraction import CONFIRMATION_THRESHOLD
from app.services.rules.types import Extraction

SUGGESTED_FIELDS: tuple[str, ...] = ("name", "net_qty_value", "net_qty_unit", "is_imported")
"""``Profile`` fields this module may propose. The names are the profile's, not the form's.

Small on purpose. These are the fields that are printed on the pack in words, which is the only
place a reading can come from.
"""

NEVER_SUGGESTED: tuple[str, ...] = (
    "surface",
    "pack_type",
    "pdp_area_cm2",
    "channel",
    "category_code",
)
"""Fields prefill must not propose, and why each one is on the list.

* ``surface`` — a physical property, not a printed word, and it selects a threshold column.
* ``pack_type`` — likewise physical: flexible, rigid, glass and can are not written on the pack.
* ``pdp_area_cm2`` — a measurement. Millimetres come from the marker homography and nowhere else
  (CLAUDE.md §3.3), and prefill runs on a downscaled photograph with no marker in play.
* ``channel`` — retail or e-commerce is a fact about the sale, not about the label.
* ``category_code`` — the client owns that vocabulary; see the module docstring.

Asserted in ``tests/test_prefill.py`` so a later "helpful" addition has to argue with a test.
"""

_QUANTITY = re.compile(r"^\s*(?P<value>\d+(?:[.,]\d+)?)\s*(?P<unit>[A-Za-z]+)\s*$")
"""A normalised net-quantity reading, split into its number and its symbol.

Anchored: this runs on ``Extraction.value``, which the pattern layer has already normalised
through the pack's unit table, so a trailing phrase means the reading is not a clean quantity and
proposing half of it would be worse than proposing nothing.
"""

_NON_ALNUM = re.compile(r"[^a-z0-9]+")


def _cleaned(value: str) -> str:
    """Lowercase, punctuation-free form for comparing a country name."""
    return _NON_ALNUM.sub(" ", value.lower()).strip()


def _is_india(value: str) -> bool:
    """Whether a country-of-origin reading names India.

    A vocabulary question, not a legal one — the same kind of question
    ``rules.types.EMBOSSED_SURFACES`` answers, and the same reason it is allowed to live in code:
    no threshold, no table row and no date is being encoded (CLAUDE.md §3.2). What follows from
    the answer is only which *suggestion* is offered to a person.
    """
    return "india" in _cleaned(value).split()


@dataclass(frozen=True)
class Suggestion:
    """One product-context field the label proposes.

    ``field`` is a ``Profile`` field name. ``value`` is always a string, including for the boolean
    ``is_imported`` ("true"), because these cross the wire as one shape and the client parses them
    into its own form types.
    """

    field: str
    value: str

    confidence: float
    """Inherited from the declaration this came from, unchanged.

    Not recomputed, and never raised: a name the model proposed carries
    ``llm_layer.LLM_CONFIDENCE`` (0.70), which is below FR-06's threshold on purpose, and it must
    still be below it after passing through here.
    """

    from_field_code: str
    """The declaration the suggestion was read from — the audit trail for "why this value"."""

    source_text: str
    """What was actually recognised, verbatim. Shown beside the field so a person is checking a
    reading against a pack rather than trusting a filled box."""

    @property
    def needs_confirmation(self) -> bool:
        """Whether FR-06's threshold puts this in front of a person before it can carry a verdict.

        Advisory here rather than binding: *every* suggestion is confirmed before a scan is
        created, because the client gates submission on it. This flags the ones the machine itself
        does not believe, so the client can say so.
        """
        return self.confidence < CONFIRMATION_THRESHOLD


def _by_code(extractions: Sequence[Extraction]) -> dict[str, Extraction]:
    """Index present declarations by field code. Blank values are not present (FR-24)."""
    return {item.field_code: item for item in extractions if item.is_present}


def _name(found: dict[str, Extraction]) -> list[Suggestion]:
    """The product name, from the common name declaration.

    Rule 6(1)(b)'s "common or generic name of the commodity" is the right source: it is what the
    thing *is* ("Iodised Salt"), which is both a usable profile name and the string a category
    search can match. The manufacturer's name is deliberately not a fallback — "Hindustan Foods
    Pvt Ltd" is a company, and putting it in the product name field would silently make the
    category search match nothing.
    """
    common = found.get("common_name")
    if common is None:
        return []

    return [
        Suggestion(
            field="name",
            value=common.value,
            confidence=common.confidence,
            from_field_code="common_name",
            source_text=common.value_raw,
        )
    ]


def _quantity(found: dict[str, Extraction]) -> list[Suggestion]:
    """Net quantity, split into a number and a prescribed symbol.

    Both halves or neither. A value with no usable unit would leave the form holding "500" with an
    empty unit box, and the unit is what selects the Rule 9 table — so a half-filled quantity is
    not a convenience, it is a field the user must now notice is wrong.
    """
    quantity = found.get("net_quantity")
    if quantity is None:
        return []

    match = _QUANTITY.match(quantity.value)
    if match is None:
        return []

    # The pack's unit table produced this symbol during normalisation. Passed through as read:
    # rewriting it here would be a second, unreviewed unit table (CLAUDE.md §3.2).
    return [
        Suggestion(
            field="net_qty_value",
            value=match.group("value").replace(",", "."),
            confidence=quantity.confidence,
            from_field_code="net_quantity",
            source_text=quantity.value_raw,
        ),
        Suggestion(
            field="net_qty_unit",
            value=match.group("unit"),
            confidence=quantity.confidence,
            from_field_code="net_quantity",
            source_text=quantity.value_raw,
        ),
    ]


def _imported(found: dict[str, Extraction]) -> list[Suggestion]:
    """Whether the label declares this an imported package.

    One direction only (rule 2 in the module docstring): this returns a suggestion of ``true`` or
    it returns nothing. ``false`` is never proposed, because ``is_imported=false`` switches the
    importer and country-of-origin rules off, and the evidence for switching a rule off cannot be
    "the label did not mention it".

    Two positive readings:

    * an **importer declaration** — Rule 6's "name and address of the importer" is what an
      imported package carries, so its presence is the direct evidence; and
    * a **country of origin that is not India** — which appears on imported packages under Rule 6
      and on e-commerce listings generally, so "India" here is not evidence of import and is
      treated as no evidence at all rather than as evidence of the opposite.
    """
    importer = found.get("importer_name")
    if importer is not None:
        return [
            Suggestion(
                field="is_imported",
                value="true",
                confidence=importer.confidence,
                from_field_code="importer_name",
                source_text=importer.value_raw,
            )
        ]

    origin = found.get("country_of_origin")
    if origin is not None and not _is_india(origin.value):
        return [
            Suggestion(
                field="is_imported",
                value="true",
                confidence=origin.confidence,
                from_field_code="country_of_origin",
                source_text=origin.value_raw,
            )
        ]

    return []


def suggest(extractions: Sequence[Extraction]) -> list[Suggestion]:
    """Propose product-context values from declarations read off a label.

    Args:
        extractions: what the extraction layer found. Only present declarations are read; a field
            extracted as an empty string is an absent declaration, not a value (FR-24).

    Returns:
        Zero or more suggestions, ordered by ``SUGGESTED_FIELDS`` so the output is deterministic
        for the same input. A field is proposed at most once, and never a field in
        ``NEVER_SUGGESTED``.
    """
    found = _by_code(extractions)

    proposed = [*_name(found), *_quantity(found), *_imported(found)]

    order = {field: index for index, field in enumerate(SUGGESTED_FIELDS)}
    return sorted(proposed, key=lambda item: order[item.field])


__all__ = ["NEVER_SUGGESTED", "SUGGESTED_FIELDS", "Suggestion", "suggest"]
